# 42 — All-categorical occurrence-frequency forest

**Decision:** retain the promoted vote. The fixed frequency-forest representation bag reaches 81.7656%, only 0.0253 percentage points above the leader.

In [1]:
from pathlib import Path
import json
import pandas as pd
PROJECT_DIR = Path('../..').resolve()
OUTPUT_DIR = PROJECT_DIR / '.runtime' / 'categorical-frequency-forest-screen'
summary = pd.read_csv(OUTPUT_DIR / 'candidate-summary.csv', index_col='candidate')
diversity = pd.read_csv(OUTPUT_DIR / 'diversity.csv')
result = json.loads((OUTPUT_DIR / 'result.json').read_text(encoding='utf-8'))
result

{'engineered_features': 43,
 'transformed_features_fold_1': 48,
 'frequency_forest_accuracy': 0.8061026936026936,
 'forest_bag_accuracy': 0.8176557239057238,
 'change_vs_promoted_identity_vote': 0.0002525252525251709,
 'passes_gate': False,
 'local_test_opened': False,
 'competition_predictions_generated': False,
 'external_evidence': 'https://github.com/drivendataorg/pump-it-up/tree/master/madRid'}

## Course-aligned lifecycle

1. **Define the goal and scope** — test a new representation from the official solution archive.
2. **Gather the data** — reuse supplied predictors and cached promoted components.
3. **Explore the data** — distinguish it from the earlier six-identity count test.
4. **Clean and preprocess the data** — preserve accepted numeric cleaning and normalise categories.
5. **Select and engineer features** — replace 29 categorical/date values with fold-fitted support counts.
6. **Define the machine-learning task** — unchanged three-class accuracy.
7. **Partition the data** — fit mappings only on each outer-training fold.
8. **Select and train candidate methods** — fit one fixed Random Forest and one equal representation bag.
9. **Evaluate and interpret the results** — compare accuracy, stability, recall and diversity.
10. **Deploy and iterate** — reject promotion and continue the external audit.

## Results

In [2]:
summary

,mean_accuracy,leader_change,fold_wins_vs_leader,worst_fold_change,repair_recall_change,passes_gate
candidate,,,,,,
forest_representation_bag_vote,0.817656,0.000253,3,-0.000842,-0.006372,False
promoted_identity_vote,0.817403,0.000000,0,0.000000,0.000000,False
frequency_replacement_vote,0.817066,-0.000337,2,-0.002946,-0.010134,False
frequency_random_forest,0.806103,-0.011301,0,-0.014625,0.017947,False
accepted_random_forest,0.805913,-0.011490,0,-0.012311,0.029236,False


In [3]:
diversity

,left_model,right_model,disagreement,left_only_correct,right_only_correct,both_wrong
0,Random Forest [baseline; frozen stratified folds],Random Forest [all categorical occurrence counts],0.064646,0.029503,0.029693,0.164394
1,Random Forest [baseline; frozen stratified folds],Promoted complete-identity vote,0.060880,0.021907,0.033396,0.160690
2,Random Forest [all categorical occurrence counts],Promoted complete-identity vote,0.071864,0.027146,0.038447,0.155450


## Interpretation

The two forests disagree on 6.46% of rows and the frequency-only forest is marginally stronger, so the representation is real rather than cosmetic. Its equal within-family bag gains only 0.025 points, well below the promotion gate; blend-weight tuning is therefore not justified.